pkg/odo/cli/url/create.go

```diff
@@ -284,8 +337,59 @@ func (o *URLCreateOptions) Run() (err error) {
 					break
 				}
 			}
+			err = o.EnvSpecificInfo.SetConfiguration("url", envinfo.EnvInfoURL{Name: o.urlName, Port: o.componentPort, ExposedPort: o.exposedPort, Kind: o.urlType})
+		} else {
+			containerEndpointMap, err := adapterutils.GetContainerEndpoints(o.devObj.Data)
+			if err != nil {
+				return errors.Wrap(err, "failed to get container endpoint map")
+			}
+			var firstContainer string
+			for containerName, endpointMap := range containerEndpointMap {
+				// save the first container name to store the new endpoint entry
+				if firstContainer == "" {
+					firstContainer = containerName
+				}
+				if _, exist := endpointMap[o.urlName]; exist {
+					return fmt.Errorf("url %v already exist in devfile endpoint entry under container %v", o.urlName, containerName)
+				}
+				for _, endpoint := range endpointMap {
+					if int(endpoint.TargetPort) == o.componentPort {
+						// devfile cannot have two endpoints within different containers share the same targetport
+						// it is because containers in a single pod shares the network namespace
+						if len(o.container) > 0 && o.container != containerName {
+							return fmt.Errorf("cannot set URL %v under container %v, TargetPort %v is being used for endpoint %v under container %v", o.urlName, o.container, o.componentPort, endpoint.Name, containerName)
+						}
+						o.container = containerName
+						break
+					}
+				}
+			}
+			if reflect.DeepEqual(containerEndpointMap, map[string]map[string]common.Endpoint{}) {
+				// devfile have no containers with endpoints
+				// pick the first container to store the new enpoint
+				o.container = o.devfileContainers[0].Name
+			}
+			if len(o.container) == 0 {
+				o.container = firstContainer
+			}
```

Comment:
I think this operation can be moved into the Complete() function and we can use the loop in https://github.com/openshift/odo/pull/3857/files#diff-476c0d834fc648caad7c721d6b10694bR207 to perform the validation by using the container ports.

Reply:
updated